go/vt/vtgate/engine/insert.go

```diff
@@ -0,0 +1,555 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/youtube/vitess/go/jsonutil"
+	"github.com/youtube/vitess/go/sqltypes"
+	"github.com/youtube/vitess/go/vt/sqlannotation"
+	"github.com/youtube/vitess/go/vt/sqlparser"
+	"github.com/youtube/vitess/go/vt/vterrors"
+	"github.com/youtube/vitess/go/vt/vtgate/vindexes"
+
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+	vtrpcpb "github.com/youtube/vitess/go/vt/proto/vtrpc"
+)
+
+var _ Primitive = (*Insert)(nil)
+
+// Insert represents the instructions to perform an insert operation.
+type Insert struct {
+	// Opcode is the execution opcode.
+	Opcode InsertOpcode
+
+	// Keyspace specifies the keyspace to send the query to.
+	Keyspace *vindexes.Keyspace
+
+	// Query specifies the query to be executed.
+	// InsertSharded plans, this value is unused.
+	// Prefix, Mid and Suffix are used instead.
+	Query string
+
+	// Values specifies values for all the vindex columns.
+	// This is a three-dimensonal data structure:
+	// Insert.Values[i] represents the values for a colvindex: len(Insert.Table.ColumnVindexes)
+	// Insert.Values[i].Values[j] represents values for each column of the colVindex: len(colVindex.Columns)
+	// Insert.Values[i].Values[j].Values[k] represents the value pulled from each row for that column: len(ins.rows)
+	Values []sqltypes.PlanValue
```

Comment:
Although I applaud the intent of this comment, it's still a bit confusingly written.

IMO something more like this would be clearer:
```
// This is a three-dimensonal data structure:
// Insert.Values[i] represents the values to be inserted for the i'th colvindex (i < len(Insert.Table.ColumnVindexes))
// Insert.Values[i].Values[j] represents values for the j'th column of the given colVindex (j < len(colVindex[i].Columns) 
// Insert.Values[i].Values[j].Values[k] represents the value pulled from row k for that column: (k < len(ins.rows))
```

Reply:
Done.